lib/sycamore/sycamore/utils/pdf.py

```diff
@@ -0,0 +1,115 @@
+import logging
+
+from io import BytesIO
+from PIL import Image
+from queue import Queue
+from subprocess import PIPE, Popen
+from threading import Thread
+from typing import List, Generator
+
+from sycamore.utils.time_trace import LogTime
+
+
+def convert_from_path_streamed(pdf_path: str) -> Generator[Image.Image, None, None]:
+    class StdoutEOF:
+        pass
+
+    class StderrEOF:
+        pass
+
+    def capture_exception(q, fn, finish_msg):
+        try:
+            fn()
+        except e:
+            q.put(e)
+
+        q.put(finish_msg)
+
+    def read_stdout(fh, q):
+        HEADER_BYTES = 40
+        need_bytes = HEADER_BYTES
+        data = b""
+        while True:
+            if need_bytes > len(data):
+                logging.debug(f"reading. have {len(data)}/{need_bytes}")
+                part = fh.read(need_bytes - len(data))
+                if part == b"":  # Eof
+                    assert len(data) == 0  # nothing left
+                    break
+                data = data + part
+            else:
+                logging.debug(f"no reading. have {len(data)}/{need_bytes}")
+
+            if len(data) < need_bytes:
+                continue
+
+            code, size, rgb = tuple(data[0:HEADER_BYTES].split(b"\n")[0:3])
+            size_x, size_y = tuple(size.split(b" "))
+            file_size = len(code) + len(size) + len(rgb) + 3 + int(size_x) * int(size_y) * 3
```

Comment:
Why not assign to need_bytes here.  Then the logic below matches above: if len(data) < need_bytes: continue.  I don't think file_size is needed.

Reply:
Done.